src/esp_loader.ts

```diff
@@ -768,6 +771,281 @@ export class ESPLoader extends EventTarget {
+  }
+  async waitDone(spiCmdReg: number, spiCmdUsr: number) {
+    for (let i = 0; i < 10; i++) {
+      let cmdValue = await this.readRegister(spiCmdReg);
+      if ((cmdValue & spiCmdUsr) == 0) {
+        return;
+      }
+    }
+    throw Error("SPI command did not complete in time");
+  }
+
+  async runSpiFlashCommand(
+    spiflashCommand: number,
+    data: number[],
+    readBits = 0
+  ) {
+    // Run an arbitrary SPI flash command.
+
+    // This function uses the "USR_COMMAND" functionality in the ESP
+    // SPI hardware, rather than the precanned commands supported by
+    // hardware. So the value of spiflash_command is an actual command
+    // byte, sent over the wire.
+
+    // After writing command byte, writes 'data' to MOSI and then
+    // reads back 'read_bits' of reply on MISO. Result is a number.
+
+    // SPI_USR register flags
+    let SPI_USR_COMMAND = 1 << 31;
+    let SPI_USR_MISO = 1 << 28;
+    let SPI_USR_MOSI = 1 << 27;
+
+    // SPI registers, base address differs ESP32* vs 8266
+    let spiAddresses = getSpiFlashAddresses(this.getChipFamily());
+    let base = spiAddresses.regBase;
+    let SPI_CMD_REG = base + 0x00;
+    let SPI_USR_REG = base + spiAddresses.usrOffs;
+    let SPI_USR2_REG = base + spiAddresses.usr2Offs;
+    let SPI_W0_REG = base + spiAddresses.w0Offs;
+
+    // SPI peripheral "command" bitmasks for SPI_CMD_REG
+    let SPI_CMD_USR = 1 << 18;
+
+    // shift values
+    let SPI_USR2_COMMAND_LEN_SHIFT = 28;
+
+    if (readBits > 32) {
+      throw new Error(
+        "Reading more than 32 bits back from a SPI flash operation is unsupported"
+      );
+    }
+    if (data.length > 64) {
+      throw new Error(
+        "Writing more than 64 bytes of data with one SPI command is unsupported"
+      );
+    }
+
+    let dataBits = data.length * 8;
+    let oldSpiUsr = await this.readRegister(SPI_USR_REG);
+    let oldSpiUsr2 = await this.readRegister(SPI_USR2_REG);
+
+    var flags = SPI_USR_COMMAND;
+
+    if (readBits > 0) {
+      flags |= SPI_USR_MISO;
+    }
+    if (dataBits > 0) {
+      flags |= SPI_USR_MOSI;
+    }
+    flags = flags;
+
+    await this.setDataLengths(spiAddresses, dataBits, readBits);
+
+    await this.writeRegister(SPI_USR_REG, flags);
+    await this.writeRegister(
+      SPI_USR2_REG,
+      (7 << SPI_USR2_COMMAND_LEN_SHIFT) | spiflashCommand
+    );
+    if (dataBits == 0) {
+      await this.writeRegister(SPI_W0_REG, 0); // clear data register before we read it
+    } else {
+      data.concat(new Array(data.length % 4).fill(0x00)); // pad to 32-bit multiple
+
+      let words = unpack("I".repeat(Math.floor(data.length / 4)), data);
+      let nextReg = SPI_W0_REG;
+
+      this.logger.debug(`Words Length: ${words.length}`);
+
+      words.forEach(async (word) => {
+        this.logger.debug(
+          `Writing word ${toHex(word)} to register offset ${toHex(nextReg)}`
+        );
+        await this.writeRegister(nextReg, word);
+        nextReg += 4;
+      });
```

Comment:
This is not working as expected. The forEach is not waiting until it's done.
```suggestion
      for (const word of words) {
        this.logger.debug(
          `Writing word ${toHex(word)} to register offset ${toHex(nextReg)}`
        );
        await this.writeRegister(nextReg, word);
        nextReg += 4;
      };
```